pkg/controller/virtualmachineimage/virtualmachineimage_controller.go

```diff
@@ -126,5 +113,32 @@ func (r *ReconcileVirtualMachineImage) validateVirtualMachineImageSpec() error {
 	if !found {
 		return goerrors.New("storage request in pvc is missing")
 	}
+	if r.vmi.Spec.Source.HTTP != "" || r.vmi.Spec.Source.Volume != nil {
+		// check node name is not empty when source is hostPath volume
+		if r.vmi.Spec.Source.Volume != nil && r.vmi.Spec.Source.Volume.HostPath != nil && r.vmi.Spec.NodeName == "" {
+			return goerrors.New("when the source is hostPath volume, the node name must not be empty")
+		}
+	} else {
+		return goerrors.New("no source set for vmim")
+	}
+	return nil
+}
+
+func (r *ReconcileVirtualMachineImage) reconcileVirtualMachineImage() error {
+	// If the pvc doesn't exist, create a pvc and update vmim's status to creating
+	if err := r.syncPvc(); err != nil {
+		return err
+	}
+	if r.vmi.Spec.Source.HTTP != "" {
```

Comment:
I think it's better to create a function like `getSourceType()` to retrieve source information, similar to the [VMVE code](https://github.com/tmax-cloud/kubevirt-image-service/blob/s3-export/pkg/controller/virtualmachinevolumeexport/virtualmachinevolumeexport_controller.go#L129).

Reply:
done